clang/lib/Tooling/DependencyScanning/IncludeTreeActionController.cpp

```diff
@@ -545,13 +562,82 @@ IncludeTreeBuilder::finishIncludeTree(CompilerInstance &ScanInstance,
       getCASTreeForFileIncludes(IncludeStack.pop_back_val());
   if (!MainIncludeTree)
     return MainIncludeTree.takeError();
-  auto FileList = cas::IncludeTree::FileList::create(DB, IncludedFiles);
+
+  if (!ScanInstance.getLangOpts().CurrentModule.empty()) {
+    SmallVector<cas::ObjectRef> Modules;
+    auto AddModule = [&](Module *M) -> llvm::Error {
+      Expected<cas::IncludeTree::Module> Mod = getIncludeTreeModule(DB, M);
+      if (!Mod)
+        return Mod.takeError();
+      Modules.push_back(Mod->getRef());
+      return Error::success();
+    };
+    if (Module *M = ScanInstance.getPreprocessor().getCurrentModule()) {
+      if (Error E = AddModule(M))
+        return std::move(E);
+    } else {
+      // When building a TU or PCH, we can have headers files that are part of
+      // both the public and private modules that are included textually. In
+      // that case we need both of those modules.
+      ModuleMap &MMap =
+          ScanInstance.getPreprocessor().getHeaderSearchInfo().getModuleMap();
+      M = MMap.findModule(ScanInstance.getLangOpts().CurrentModule);
+      assert(M && "missing current module?");
+      if (Error E = AddModule(M))
+        return std::move(E);
+      Module *PM =
+          MMap.findModule(ScanInstance.getLangOpts().ModuleName + "_Private");
+      if (PM)
+        if (Error E = AddModule(PM))
+          return std::move(E);
+    }
+
+    auto ModMap = cas::IncludeTree::ModuleMap::create(DB, Modules);
+    if (!ModMap)
+      return ModMap.takeError();
+    ModuleMapRef = ModMap->getRef();
+  }
+
+  auto FileList =
+      cas::IncludeTree::FileList::create(DB, IncludedFiles.getArrayRef());
   if (!FileList)
     return FileList.takeError();
 
   return cas::IncludeTreeRoot::create(DB, MainIncludeTree->getRef(),
-                                      FileList->getRef(), PCHRef,
-                                      ModuleMapFileRef);
+                                      FileList->getRef(), PCHRef, ModuleMapRef);
+}
+
+Error IncludeTreeBuilder::addModuleInputs(ASTReader &Reader) {
+  for (serialization::ModuleFile &MF : Reader.getModuleManager()) {
+    // Only add direct imports to avoid duplication. Each include tree is a
+    // superset of its imported modules' include trees.
+    if (!MF.isDirectlyImported())
+      continue;
+
+    assert(!MF.IncludeTreeID.empty() && "missing include-tree for import");
+
+    std::optional<cas::CASID> ID;
+    if (Error E = DB.parseID(MF.IncludeTreeID).moveInto(ID))
+      return E;
+    std::optional<cas::ObjectRef> Ref = DB.getReference(*ID);
+    if (!Ref)
+      return DB.createUnknownObjectError(*ID);
+    std::optional<cas::IncludeTreeRoot> Root;
+    if (Error E = cas::IncludeTreeRoot::get(DB, *Ref).moveInto(Root))
+      return E;
+    std::optional<cas::IncludeTree::FileList> Files;
+    if (Error E = Root->getFileList().moveInto(Files))
+      return E;
+
+    Error E = Files->forEachFile([&](auto IF, auto Size) -> Error {
```

Comment:
Interesting idea, so the advantage is fewer CAS references created and moving the cost of deduplicating the files to compile time instead of scan time.

I'll take a look 👍

Reply:
"fewer CAS references created" and we are not copying the `FileList`s to larger `FileList`s.

Reply:
Implemented in d67f7857d04c277.